### PR TITLE
Implement AdaGrad #127

### PR DESCRIPTION
**Pull request by Mistobaan**

### CODE REVIEW

File: Sources/TensorFlow/Optimizer.swift [hunk 1]
```diff
@@ -294,3 +294,62 @@ public class RiemannSGD<Model: Layer, Scalar: FloatingPoint>: Optimizer
...
+/// the square root of the sum of all the historical saquared values of the gradient.
+/// 
+/// Reference: ["Adaptive Subgradient Methods for Online Learning and Stochastic Optimization"](
+///  http://www.jmlr.org/papers/volume12/duchi11a/duchi11a.pdf)
+/// 
+public class AdaGrad<Model: Layer>: Optimizer
+    where Model.AllDifferentiableVariables == Model.TangentVector {
+    /// The learning rate.
+    public var learningRate: Float
+    // Rho is the smoothing factor. Typical values are 0.5, 0.9, 0.99. (smooth over 2, 10, 100 examples)
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: * Use 3 slashes for doc comments.
* Add backquotes around numeric values in comments. 

```suggestion
    /// The smoothing factor (ρ). Typical values are `0.5`, `0.9`, and `0.99`, for smoothing over 2,
    /// 10, and 100 examples, respectively).
```

File: Sources/TensorFlow/Optimizer.swift [hunk 2]
```diff
@@ -294,3 +294,62 @@ public class RiemannSGD<Model: Layer, Scalar: FloatingPoint>: Optimizer
...
+
+        alpha = model.allDifferentiableVariables
+        for kp in alpha.recursivelyAllWritableKeyPaths(to: Tensor<Float>.self) {
+            alpha[keyPath: kp].resetToZero()
+        }
+        for kp in alpha.recursivelyAllWritableKeyPaths(to: Tensor<Double>.self) {
+            alpha[keyPath: kp].resetToZero()
+        }
+    }
+
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: Remove one extra line.

File: Sources/TensorFlow/Optimizer.swift [hunk 3]
```diff
@@ -294,3 +294,62 @@ public class RiemannSGD<Model: Layer, Scalar: FloatingPoint>: Optimizer
         model = model.moved(along: learningRate * (.zero - direction))
     }
 }
+
+/// AdaGrad optimizer.
+/// Individually adapts the learning rates of  all model parameters by scaling them inversely proportional to
```
Comment (saeta) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
/// Individually adapts the learning rates of all model parameters by scaling them inversely proportional to 
```

File: Sources/TensorFlow/Optimizer.swift [hunk 4]
```diff
@@ -294,3 +294,62 @@ public class RiemannSGD<Model: Layer, Scalar: FloatingPoint>: Optimizer
         model = model.moved(along: learningRate * (.zero - direction))
     }
 }
+
+/// AdaGrad optimizer.
+/// Individually adapts the learning rates of  all model parameters by scaling them inversely proportional to 
+/// the square root of the sum of all the historical saquared values of the gradient.
```
Comment (saeta) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
/// the square root of the sum of all the historical squared values of the gradient.
```

File: Sources/TensorFlow/Optimizer.swift [hunk 5]
```diff
@@ -294,3 +294,62 @@ public class RiemannSGD<Model: Layer, Scalar: FloatingPoint>: Optimizer
         model = model.moved(along: learningRate * (.zero - direction))
     }
 }
+
+/// AdaGrad optimizer.
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
/// AdaGrad optimizer.
///
```

File: Sources/TensorFlow/Optimizer.swift (hunk 3 above)
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
/// Individually adapts the learning rates of all model parameters by scaling them inversely proportional to 
```

File: Sources/TensorFlow/Optimizer.swift [hunk 6]
```diff
@@ -294,3 +294,62 @@ public class RiemannSGD<Model: Layer, Scalar: FloatingPoint>: Optimizer
...
+        }
+        for kp in alpha.recursivelyAllWritableKeyPaths(to: Tensor<Double>.self) {
+            alpha[keyPath: kp].resetToZero()
+        }
+    }
+
+
+    public func update(_ model: inout Model.AllDifferentiableVariables,
+                       along direction: Model.TangentVector) {
+        for kp in model.recursivelyAllWritableKeyPaths(to: Tensor<Float>.self) {
```
Comment (saeta) [inline Sources/TensorFlow/Optimizer.swift]: Could you add some comments describing what each of the for loops are doing to help the reader?

File: Sources/TensorFlow/Optimizer.swift [hunk 7]
```diff
@@ -294,3 +294,62 @@ public class RiemannSGD<Model: Layer, Scalar: FloatingPoint>: Optimizer
...
+            alpha[keyPath: kp] = rho + pow(direction[keyPath: kp], 2)
```
Comment (eaplatanios) [inline Sources/TensorFlow/Optimizer.swift]: This is a minor comment, but changing this to `direction[keyPath: kp].squared()` should result in better performance.

File: Sources/TensorFlow/Optimizer.swift [hunk 8]
```diff
@@ -294,3 +294,62 @@ public class RiemannSGD<Model: Layer, Scalar: FloatingPoint>: Optimizer
...
+            model[keyPath: kp] -=
+                learningRate * direction[keyPath: kp] / (sqrt(alpha[keyPath: kp] + epsilon))
+        }
+        for kp in model.recursivelyAllWritableKeyPaths(to: Tensor<Double>.self) {
+            alpha[keyPath: kp] =
+                Double(rho) + pow(direction[keyPath: kp], 2)
```
Comment (eaplatanios) [inline Sources/TensorFlow/Optimizer.swift]: Same comment as above.

Comment (eaplatanios) [reply Sources/TensorFlow/Optimizer.swift]: Also, I believe this fits in one line.

File: Sources/TensorFlow/Optimizer.swift (hunk 7 above)
Comment (Mistobaan) [reply Sources/TensorFlow/Optimizer.swift]: Good to know. The other implementations could benefit from this version too.

**Comment by Shashi456** [comment]
@rxwei @saeta @eaplatanios As far as my understanding of Adagrad goes,  This implementation is missing a learning rate decay.  It should also have a precondition and the learning rate should get updated each loop. 

One of the features of the Adagrad algorithm is the learning rate decay, since the authors argue that in usual Machine learning, learning rate remained constant.

**Comment by Mistobaan** [comment]
Yes @Shashi456, there is no learning rate decay. I think embedding the learning rate decay strategy into the optimization logic is not good. There are many different ways to update the learning rate (see: https://arxiv.org/abs/1506.01186).

**Comment by Shashi456** [comment]
@Mistobaan I'll take a look. A variety of implementations i've looked across frameworks seem to have some kind of decay in place. Thanks for the paper link :P

**Comment by eaplatanios** [comment]
Learning rate decay is not part of adagrad per se and is independent.
Adagrad simply adapts the learning rate by using the accumulated gradient
norms and that’s sufficient to obtain the guarantees shown in the paper as
far as I remember. There is many kinds of learning rate schedules you can
use on top of adagrad but I don’t believe they should be mixed up with the
optimizer implementations as they can be used with any optimizer.

**Comment by Mistobaan** [comment]
Updates with all the comments. Added simple XOR test case.

File: Tests/TensorFlowTests/OptimizersTests.swift [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import XCTest
+@testable import TensorFlow
+
+final class OptimizerTests: XCTestCase {
+
```
Comment (rxwei) [inline Tests/TensorFlowTests/OptimizersTests.swift]: Remove extra line.
```suggestion
```

File: Tests/TensorFlowTests/OptimizersTests.swift [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    func testAdaGrad() {
+        struct Classifier: Layer {
+            var l1, l2: Dense<Float>
+            init(hiddenSize: Int) {
+                l1 = Dense<Float>(
+                        inputSize: 2,
```
Comment (rxwei) [inline Tests/TensorFlowTests/OptimizersTests.swift]:
```suggestion
                    inputSize: 2,
```

Indent all line breaks by 4 whitespaces. Same for other lines.

File: Tests/TensorFlowTests/OptimizersTests.swift [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+                return meanSquaredError(predicted: ŷ, expected: y)
+            }
+            optimizer.update(&classifier.allDifferentiableVariables, along: 𝛁model)
+        }
+
+        // trained classifier should return valid values
+        ŷ = classifier.inferring(from: x)
+        XCTAssertEqual(round(ŷ), y)
+    }
+
```
Comment (rxwei) [inline Tests/TensorFlowTests/OptimizersTests.swift]: Remove extra line.
